components/brave_wallet/browser/internal/hd_key_sr25519_unittest.cc

```diff
@@ -0,0 +1,204 @@
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_wallet/browser/internal/hd_key_sr25519.h"
+
+#include <vector>
+
+#include "base/strings/string_number_conversions.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace brave_wallet {
+
+inline constexpr uint8_t kSchnorrkelSeed[] = {
+    157, 97,  177, 157, 239, 253, 90, 96,  186, 132, 74,
+    244, 146, 236, 44,  196, 68,  73, 197, 105, 123, 50,
+    105, 25,  112, 59,  172, 3,   28, 174, 127, 96,
+};
+
+TEST(HDKeySr25519, GenerateFromSeed) {
+  auto kpresult = HDKeySr25519::GenerateFromSeed({});
+  EXPECT_FALSE(kpresult);
+
+  kpresult = HDKeySr25519::GenerateFromSeed({1, 2, 3, 4});
+  EXPECT_FALSE(kpresult);
+
+  kpresult = HDKeySr25519::GenerateFromSeed(kSchnorrkelSeed);
+  EXPECT_TRUE(kpresult);
+
+  kpresult = HDKeySr25519::GenerateFromSeed(std::array<uint8_t, 31>{});
+  EXPECT_FALSE(kpresult);
+
+  kpresult = HDKeySr25519::GenerateFromSeed(std::array<uint8_t, 33>{});
+  EXPECT_FALSE(kpresult);
+}
+
+TEST(HDKeySr25519, GetPublicKey) {
+  auto kpresult = HDKeySr25519::GenerateFromSeed(kSchnorrkelSeed);
+  ASSERT_TRUE(kpresult);
+
+  auto public_key = kpresult->GetPublicKey();
+  EXPECT_EQ(public_key.size(), std::size_t{32});
+
+  auto kpresult2 = HDKeySr25519::GenerateFromSeed(kSchnorrkelSeed);
+  ASSERT_TRUE(kpresult2);
+
+  EXPECT_EQ(kpresult->GetPublicKey(), kpresult2->GetPublicKey());
+}
+
+TEST(HDKeySr25519, SignAndVerify) {
+  auto kpresult = HDKeySr25519::GenerateFromSeed(kSchnorrkelSeed);
+  ASSERT_TRUE(kpresult);
+
+  auto public_key = kpresult->GetPublicKey();
+  EXPECT_EQ(public_key.size(), std::size_t{32});
+
+  unsigned char const message[] = {1, 2, 3, 4, 5, 6};
+  auto sig = kpresult->SignMessage(message);
+
+  auto is_verified = kpresult->VerifyMessage(sig, message);
+  EXPECT_TRUE(is_verified);
+
+  std::array<uint8_t, 64> bad_sig = {};
+  is_verified = kpresult->VerifyMessage(bad_sig, message);
+  EXPECT_FALSE(is_verified);
+
+  std::array<uint8_t, 64> bad_message = {};
+  is_verified = kpresult->VerifyMessage(sig, bad_message);
+  EXPECT_FALSE(is_verified);
+}
+
+TEST(HDKeySr25519, HardDerive) {
+  auto kpresult = HDKeySr25519::GenerateFromSeed(kSchnorrkelSeed);
+  ASSERT_TRUE(kpresult);
+
+  // manually create a SCALE-encoded chaincode value
+  unsigned char path1[] = {20, 'A', 'l', 'i', 'c', 'e'};
+  unsigned char path2[] = {20, 'e', 'c', 'i', 'l', 'A'};
+
+  auto derived1 = kpresult->DeriveHard(path1);
+  auto derived2 = kpresult->DeriveHard(path2);
+  auto derived3 = kpresult->DeriveHard(path1);
+
+  EXPECT_EQ(derived1->GetPublicKey(), derived3->GetPublicKey());
+  EXPECT_NE(derived1->GetPublicKey(), derived2->GetPublicKey());
+
+  auto kpresult2 = HDKeySr25519::GenerateFromSeed(
+      {250, 199, 149, 157, 191, 231, 47,  5,   46,  90,  12,
+       60,  141, 101, 48,  242, 2,   176, 47,  216, 249, 245,
+       202, 53,  128, 236, 141, 235, 119, 151, 71,  158});
```

Comment:
I would advise to wrap this in `std::to_array` here and in other places doing the same, because I think there are cases where this is not guaranteed to be ro data when instantiating like that.

```cxx
  auto kpresult2 = HDKeySr25519::GenerateFromSeed(
      std::to_array<uint8_t>({250, 199, 149, 157, 191, 231, 47,  5,   46,  90,  12,
       60,  141, 101, 48,  242, 2,   176, 47,  216, 249, 245,
       202, 53,  128, 236, 141, 235, 119, 151, 71,  158}));
```

Reply:
Actually, never mind, as I realise this gets digested in that function.